sdk/client/src/client.rs

```diff
@@ -91,13 +91,10 @@ impl Client {
                 .await
                 .map_err(WaitForTransactionError::GetTransactionError)?;
             if let (Some(txn), state) = txn_resp.into_parts() {
-                if txn.hash.to_hex() != txn_hash {
+                if txn.hash != txn_hash {
                     return Err(WaitForTransactionError::TransactionHashMismatchError(txn));
                 }
-                match txn.vm_status {
-                    diem_json_rpc_types::views::VMStatusView::Executed => {}
-                    _ => return Err(WaitForTransactionError::TransactionExecutionFailed(txn)),
-                }
```

Comment:
I think it is too easy to make mistake for a client calls to wait for txn, and use the result without checking the txn is executed successfully. Thus checking the execution result and return proper error is better. We can see all caller need handle it.
It is better for test to assert the result, but I think assert error is WaitForTransactionError::TransactionExecutionFailed is same.

Reply:
I don't know if I really agree. This function is named "wait_for_transaction" so I would expect this to wait for a txn to be committed and then give me back that txn, failure or no. Really what we probably want is another fn that is "wait_for_executed_transaction" or something that has the extra check in it. Basically having the lowest building block be more flexible while allowing for building more complex things on top.

Reply:
I don't disagree. Internally, we can split it into smaller building blocks, but a high level one function waiting for execution succeed transaction is useful in practice.

Reply:
just spit balling a bit, but another approach might be `wait_for_transaction(..) -> Result<Result<Response<TransactionView>, ExecutionError>, ClientError>`. that would help communicate the different, orthogonal error modes.

Reply:
It looks like most uses in our code base are assuming that `wait_for_signed_transaction` checks for if the txn executed properly. So I'm going to leave this functionality for that function but will instead change the lower level building block `wait_for_transaction` to not error on a txn that failed. I feel like that's probably a good compromise for now